publisher-sdk/src/test/java/com/criteo/publisher/LiveCdbCallListenerTests.kt

```diff
@@ -0,0 +1,235 @@
+    whenever(bidManager.isBidSilent(cachedCdbResponseSlot)).thenReturn(false)
+    whenever(bidManager.consumeCachedBid(cacheAdUnit)).thenReturn(cachedCdbResponseSlot)
+
+    liveCdbCallListener.onTimeBudgetExceeded()
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).consumeCachedBid(cacheAdUnit)
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidManager).setCacheAdUnits(cdbResponse.slots)
+    verify(bidListener, never()).onBidResponse(cachedCdbResponseSlot)
+    verify(bidListener, never()).onBidResponse(freshCdbResponseSlot)
+    verify(bidListener).onNoBid()
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+    verify(bidLifecycleListener).onBidConsumed(any(), any())
+  }
+
+  @Test
+  fun onBidResponse_givenTimeBudgetExceededAndCacheEntryExpiredAndSilent_ThenCacheNewResponseAndCallNoBid() {
+    whenever(cdbResponse.slots).thenReturn(listOf(freshCdbResponseSlot))
+    whenever(cdbResponse.timeToNextCall).thenReturn(1_000)
+    whenever(bidManager.consumeCachedBid(cacheAdUnit)).thenReturn(null)
+    whenever(bidManager.hasBidExpired(cachedCdbResponseSlot)).thenReturn(true);
+    whenever(bidManager.isBidSilent(cachedCdbResponseSlot)).thenReturn(false)
+    whenever(bidManager.consumeCachedBid(cacheAdUnit)).thenReturn(cachedCdbResponseSlot)
+
+    liveCdbCallListener.onTimeBudgetExceeded()
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).consumeCachedBid(cacheAdUnit)
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidManager).setCacheAdUnits(cdbResponse.slots)
+    verify(bidListener, never()).onBidResponse(cachedCdbResponseSlot)
+    verify(bidListener, never()).onBidResponse(freshCdbResponseSlot)
+    verify(bidListener).onNoBid()
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+    verify(bidLifecycleListener).onBidConsumed(any(), any())
+  }
+
+  @Test
+  fun onBidResponse_givenTimeBudgetExceededAndNoValidCacheEntry_ThenCacheNewResponseAndCallNoBid() {
+    whenever(cdbResponse.slots).thenReturn(listOf(freshCdbResponseSlot))
+    whenever(cdbResponse.timeToNextCall).thenReturn(1_000)
+    whenever(bidManager.consumeCachedBid(cacheAdUnit)).thenReturn(null)
+
+    liveCdbCallListener.onTimeBudgetExceeded()
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).consumeCachedBid(cacheAdUnit)
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidManager).setCacheAdUnits(cdbResponse.slots)
+    verify(bidListener, never()).onBidResponse(cachedCdbResponseSlot)
+    verify(bidListener, never()).onBidResponse(freshCdbResponseSlot)
+    verify(bidListener).onNoBid()
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+    verify(bidLifecycleListener, never()).onBidConsumed(any(), any())
+  }
+
+  @Test
+  fun onBidResponse_givenTimeoutAndNoValidCacheEntry_ThenDontCacheResponseAndCallNoBid() {
```

Comment:
It is a timeout, see how I am calling: `liveCdbCallListener.onCdbError(cdbRequest, exception)` on line 224.

Reply:
Ah ok, I was not assuming this. You can still have a real CDB error, even after the time budget exceeded. But yes, generally it would be a timeout.
It was really nit and just about wording to match the other test methods.